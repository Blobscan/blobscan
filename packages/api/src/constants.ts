export const DEFAULT_LIMIT = 50;
