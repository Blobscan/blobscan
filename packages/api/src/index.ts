import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server";

import type { AppRouter } from "./app-router";

export { appRouter, type AppRouter } from "./app-router";
export { createTRPCContext, createTRPCInnerContext } from "./context";
export type {
  TRPCInnerContext as TRCInnerContext,
  TRPCContext,
} from "./context";

export * from "@trpc/server/adapters/express";

export { env } from "./env";
export { gracefulShutdown } from "./graceful-shutdown";
export { metricsHandler } from "./instrumentation";
export type {
  BlobStorage,
  Rollup,
  Metric,
  MetricHistogram,
  MetricHistogramBucket,
  Metrics,
  MetricsClient,
} from "@blobscan/db";

/**
 * Inference helpers for input types
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helpers for output types
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
