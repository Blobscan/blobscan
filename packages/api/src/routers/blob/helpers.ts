import type {
  BlobscanPrismaClient,
  EthUsdPrice,
  ExtendedBlobDataStorageReferenceSelect,
  Prisma,
} from "@blobscan/db";
import { EthUsdPriceModel } from "@blobscan/db/prisma/zod";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import {
  normalizePrismaBlobDataStorageReferencesFields,
  normalizePrismaTransactionFields,
} from "../../utils";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
  prismaBlobOnTransactionSchema,
} from "../../zod-schemas";

const blobDataStorageReferenceSelect = {
  blobStorage: true,
  url: true,
} satisfies ExtendedBlobDataStorageReferenceSelect;

export const extendedBlobDataStorageReferenceArgs: Prisma.Args<
  BlobscanPrismaClient["blobDataStorageReference"],
  "findFirstOrThrow"
> = {
  select: blobDataStorageReferenceSelect,
  orderBy: {
    blobStorage: "asc",
  },
};

export const baseBlobSelect = {
  commitment: true,
  proof: true,
  size: true,
  usageSize: true,
  versionedHash: true,
} satisfies Prisma.BlobSelect;

export const baseBlobOnTransactionSelect = {
  blobHash: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  txHash: true,
  txIndex: true,
} satisfies Prisma.BlobsOnTransactionsSelect;

export type ExtendedBlobDataStorageReference = NonNullable<
  Prisma.Result<
    BlobscanPrismaClient["blobDataStorageReference"],
    { select: typeof blobDataStorageReferenceSelect },
    "findFirst"
  >
>;
type PrismaBlob = Prisma.BlobGetPayload<{
  select: typeof baseBlobSelect;
}> & {
  dataStorageReferences: ExtendedBlobDataStorageReference[];
};

type PrismaBlobOnTransaction = Prisma.BlobsOnTransactionsGetPayload<{
  select: typeof baseBlobOnTransactionSelect;
}>;

export type CompletedPrismaBlob = Prettify<
  PrismaBlob & {
    transactions: Prettify<
      PrismaBlobOnTransaction & {
        block?: ExpandedBlock;
        transaction?: ExpandedTransaction & {
          block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
        };
      }
    >[];
  }
>;

export type CompletedPrismaBlobOnTransaction = Prettify<
  PrismaBlobOnTransaction & {
    blob: {
      dataStorageReferences: ExtendedBlobDataStorageReference[];
    } & ExpandedBlob;
    block?: ExpandedBlock;
    transaction?: ExpandedTransaction & {
      block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
    };
  }
>;
export function createBlobSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobSelect,
    dataStorageReferences: extendedBlobDataStorageReferenceArgs,
    transactions: {
      select: {
        ...baseBlobOnTransactionSelect,
        ...blockExpand,
        ...txExpand,
      },
      orderBy: {
        blockNumber: "desc",
      },
    },
  } satisfies Prisma.BlobSelect;
}

export function createBlobsOnTransactionsSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobOnTransactionSelect,
    blob: {
      select: {
        ...baseBlobSelect,
        dataStorageReferences: extendedBlobDataStorageReferenceArgs,
      },
    },
    ...txExpand,
    ...blockExpand,
  } satisfies Prisma.BlobsOnTransactionsSelect;
}

const responseBlobOnTxSchema = prismaBlobOnTransactionSchema
  .omit({ blobHash: true })
  .required({
    blockHash: true,
    blockNumber: true,
    blockTimestamp: true,
    index: true,
    txHash: true,
    txIndex: true,
  });

export const responseBlobSchema = baseBlobSchema.extend({
  transactions: z.array(
    responseBlobOnTxSchema
      .merge(
        baseTransactionSchema
          .omit({
            hash: true,
            blockNumber: true,
            blockTimestamp: true,
            blockHash: true,
          })
          .partial()
      )
      .extend({
        block: baseBlockSchema
          .omit({
            hash: true,
            number: true,
            timestamp: true,
          })
          .partial({
            blobGasBaseUsdFee: true,
            blobGasUsdPrice: true,
          })
          .optional(),
        ethUsdPrice: EthUsdPriceModel.shape.price.optional(),
      })
  ),
});

export const responseBlobOnTransactionSchema = responseBlobOnTxSchema
  .merge(baseBlobSchema)
  .extend({
    block: baseBlockSchema
      .omit({
        hash: true,
        number: true,
        timestamp: true,
      })
      .optional(),
    transaction: baseTransactionSchema
      .omit({
        hash: true,
        blockNumber: true,
        blockTimestamp: true,
        blockHash: true,
      })
      .optional(),
  });

export type ResponseBlob = z.input<typeof responseBlobSchema>;
export type ResponseBlobOnTransaction = z.input<
  typeof responseBlobOnTransactionSchema
>;

export function toResponseBlob(
  {
    dataStorageReferences,
    transactions: blobOnTxs,
    ...prismaBlob
  }: CompletedPrismaBlob,
  ethUsdPrices: EthUsdPrice["price"][]
): ResponseBlob {
  const responseBlob: ResponseBlob = {
    ...prismaBlob,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    transactions: blobOnTxs.map(
      (
        {
          block: { computeUsdFields, ...restBlock } = {},
          transaction: prismaTx = {},
          ...restBlobOnTx
        },
        i
      ) => {
        const txEthUsdPrice = ethUsdPrices[i];

        const blockUsdFields =
          txEthUsdPrice && computeUsdFields
            ? computeUsdFields(txEthUsdPrice)
            : undefined;
        const txBlock = {
          ...restBlock,
          ...(blockUsdFields ?? {}),
        };

        let responseBlobOnTx: ResponseBlob["transactions"][number] = {
          ...restBlobOnTx,
          ...txBlock,
          ethUsdPrice: txEthUsdPrice?.toNumber(),
        };

        if (isFullyDefined(prismaTx)) {
          const {
            block,
            decodedFields,
            from,
            fromId,
            toId,
            computeBlobGasBaseFee,
            computeUsdFields,
            ...restPrismaTx
          } = prismaTx;

          responseBlobOnTx = {
            ...responseBlobOnTx,
            ...restPrismaTx,
            ...normalizePrismaTransactionFields({
              decodedFields,
              from,
              fromId,
              toId,
            }),
            blobGasBaseFee: computeBlobGasBaseFee(block.blobGasPrice),
            ...(txEthUsdPrice
              ? computeUsdFields({
                  blobGasPrice: block.blobGasPrice,
                  ethUsdPrice: txEthUsdPrice,
                })
              : {}),
          };
        }

        return responseBlobOnTx;
      }
    ),
  };

  return responseBlob;
}

export function toResponseBlobOnTransaction(
  prismaBlobOnTransaction: CompletedPrismaBlobOnTransaction,
  ethUsdPrice?: EthUsdPrice["price"]
): ResponseBlobOnTransaction {
  const {
    blobHash,
    blob: { dataStorageReferences, ...blob },
    transaction,
    block,
    ...restBlobOnTransaction
  } = prismaBlobOnTransaction;

  const responseBlobOnTx: ResponseBlobOnTransaction = {
    ...blob,
    versionedHash: blobHash,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    ...restBlobOnTransaction,
  };

  if (transaction) {
    const {
      block: txBlock,
      decodedFields,
      from,
      fromId,
      toId,
      computeBlobGasBaseFee,
      computeUsdFields,
      ...restTx
    } = transaction;

    responseBlobOnTx.transaction = {
      ...restTx,
      blobGasPrice: txBlock.blobGasPrice,
      ...normalizePrismaTransactionFields({
        decodedFields,
        from,
        fromId,
        toId,
      }),
      blobGasBaseFee: computeBlobGasBaseFee(txBlock.blobGasPrice),
      ...(ethUsdPrice
        ? computeUsdFields({
            blobGasPrice: txBlock.blobGasPrice,
            ethUsdPrice,
          })
        : {}),
    };
  }

  if (block) {
    const { computeUsdFields, ...restBlock } = block;
    responseBlobOnTx.block = {
      ...restBlock,
      ...(ethUsdPrice ? computeUsdFields(ethUsdPrice) : {}),
    };
  }

  return responseBlobOnTx;
}
