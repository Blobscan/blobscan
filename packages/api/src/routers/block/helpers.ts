import type { BlobStorageManager } from "@blobscan/blob-storage-manager";
import type { BlobscanPrismaClient, Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Filters } from "../../middlewares/withFilters";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import { retrieveBlobData } from "../../utils";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
} from "../../zod-schemas";

const prismaBlockSelect = {
  hash: true,
  number: true,
  timestamp: true,
  slot: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  blobGasPrice: true,
  excessBlobGas: true,
} satisfies Prisma.BlockSelect;

export type PrismaBlock = Prisma.BlockGetPayload<{
  select: typeof prismaBlockSelect;
}>;

export type CompletePrismaBlock = Prettify<
  PrismaBlock & {
    transactions: Prettify<
      {
        hash: string;
      } & Partial<ExpandedTransaction> & {
          blobs: Prettify<
            { blobHash: string } & {
              blob?: ExpandedBlob;
            }
          >[];
        }
    >[];
  }
>;

export function createBlockSelect(expands: Expands, filters?: Filters) {
  const blobExpand = expands.blob ? { blob: expands.blob } : {};
  const transactionExpand = expands.transaction?.select ?? {};

  return {
    ...prismaBlockSelect,
    transactions: {
      select: {
        hash: true,
        ...transactionExpand,
        blobs: {
          select: {
            blobHash: true,
            ...blobExpand,
          },
          orderBy: {
            index: "asc",
          },
        },
      },
      orderBy: {
        index: "asc",
      },
      where: filters?.transactionFilters,
    },
  } satisfies Prisma.BlockSelect;
}

export const responseBlockSchema = baseBlockSchema.extend({
  transactions: z.array(
    baseTransactionSchema
      .omit({
        blockHash: true,
        blockNumber: true,
        blockTimestamp: true,
      })
      .partial()
      .required({ hash: true })
      .extend({
        blobs: z.array(
          baseBlobSchema.partial().required({ versionedHash: true })
        ),
      })
  ),
});

export type ResponseBlock = z.input<typeof responseBlockSchema>;

export function toResponseBlock(
  prismaBlock: CompletePrismaBlock
): ResponseBlock {
  const transactions = prismaBlock.transactions.map(
    ({ hash, blobs, ...prismaTx }) => ({
      hash,
      ...(isFullyDefined(prismaTx)
        ? {
            ...prismaTx,
            ...normalizePrismaTransactionFields(prismaTx),
            ...deriveTransactionFields({
              ...prismaTx,
              blobGasPrice: prismaBlock.blobGasPrice,
            }),
          }
        : {}),
      blobs: blobs.map(({ blobHash, blob }) =>
        blob
          ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
          : { versionedHash: blobHash }
      ),
    })
  );

  return {
    ...prismaBlock,
    transactions,
  };
}

export type BlockId = "hash" | "number" | "slot";
export type BlockIdField =
  | { type: "hash"; value: string }
  | { type: "number"; value: number }
  | { type: "slot"; value: number };

function buildBlockWhereClause(
  { type, value }: BlockIdField,
  filters: Filters
): Prisma.BlockWhereInput {
  switch (type) {
    case "hash": {
      return { hash: value };
    }
    case "number": {
      return { number: value, transactionForks: filters.blockType };
    }
    case "slot": {
      return { slot: value, transactionForks: filters.blockType };
    }
  }
}

export async function fetchBlock(
  blockId: BlockIdField,
  {
    blobStorageManager,
    prisma,
    filters,
    expands,
  }: {
    blobStorageManager: BlobStorageManager;
    prisma: BlobscanPrismaClient;
    filters: Filters;
    expands: Expands;
  }
): Promise<CompletePrismaBlock | undefined> {
  const select = createBlockSelect(expands);
  const where = buildBlockWhereClause(blockId, filters);

  const prismaBlock = (await prisma.block.findFirst({
    select,
    where,
  })) as unknown as CompletePrismaBlock | null;

  if (!prismaBlock) {
    return;
  }

  if (expands.blobData) {
    await Promise.all(
      prismaBlock.transactions
        .flatMap((tx) => tx.blobs)
        .map(async ({ blobHash, blob }) => {
          const dataStorageReferences = blob?.dataStorageReferences;
          if (dataStorageReferences) {
            const data = await retrieveBlobData(blobStorageManager, {
              dataStorageReferences,
              versionedHash: blobHash,
            });

            blob.data = data;
          }
        })
    );
  }

  return prismaBlock;
}
