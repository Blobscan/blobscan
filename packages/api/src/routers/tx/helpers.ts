import type { Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseTransactionSchema,
  baseBlobSchema,
  baseBlockSchema,
} from "../../zod-schemas";

const transactionSelect = {
  hash: true,
  fromId: true,
  toId: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  gasPrice: true,
  maxFeePerBlobGas: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  decodedFields: true,
  from: {
    select: {
      address: true,
      rollup: true,
    },
  },
} satisfies Prisma.TransactionSelect;

type PrismaTransaction = Prisma.TransactionGetPayload<{
  select: typeof transactionSelect;
}>;

export type CompletePrismaTransaction = Prettify<
  PrismaTransaction & {
    decodedFields: NonNullable<PrismaTransaction["decodedFields"]>;
    block: Prettify<
      Partial<ExpandedBlock> & {
        blobGasPrice: ExpandedBlock["blobGasPrice"];
      }
    >;
    blobs: Prettify<
      { blobHash: string } & {
        blob?: ExpandedBlob;
      }
    >[];
  }
>;
export function createTransactionSelect(expands: Expands) {
  const blockExpand = expands.block?.select;
  const blobExpand = expands.blob;

  return {
    ...transactionSelect,
    block: {
      select: {
        ...(blockExpand ?? {}),
        blobGasPrice: true,
      },
    },
    blobs: {
      select: {
        blobHash: true,
        ...(blobExpand ? { blob: blobExpand } : {}),
      },
      orderBy: {
        index: "asc",
      },
    },
  } satisfies Prisma.TransactionSelect;
}

export const responseTransactionSchema = baseTransactionSchema.extend({
  block: baseBlockSchema
    .omit({
      hash: true,
      number: true,
      timestamp: true,
    })
    .partial()
    .required({
      blobGasPrice: true,
    }),
  blobs: z.array(baseBlobSchema.partial().required({ versionedHash: true })),
});

export type ResponseTransaction = z.input<typeof responseTransactionSchema>;

export function toResponseTransaction(
  prismaTx: CompletePrismaTransaction
): ResponseTransaction {
  const { blobs: blobsOnTxs, block } = prismaTx;
  const normalizedFields = normalizePrismaTransactionFields(prismaTx);
  const derivedFields = deriveTransactionFields({
    ...prismaTx,
    blobGasPrice: block.blobGasPrice,
  });
  const blobs = blobsOnTxs.map(({ blobHash, blob }) =>
    blob
      ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
      : { versionedHash: blobHash }
  );

  return {
    ...prismaTx,
    ...normalizedFields,
    ...derivedFields,
    blobs,
  };
}
