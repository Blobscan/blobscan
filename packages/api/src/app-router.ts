import { z } from "@blobscan/zod";

import { publicProcedure } from "./procedures";
import { blobRouter } from "./routers/blob";
import { blobStoragesStateRouter } from "./routers/blob-storages-state";
import { blockRouter } from "./routers/block";
import { blockchainSyncStateRouter } from "./routers/blockchain-sync-state";
import { indexerRouter } from "./routers/indexer";
import { searchRouter } from "./routers/search";
import { statsRouter } from "./routers/stats";
import { transactionRouter } from "./routers/tx";
import { t } from "./trpc-client";

export const appRouter = t.router({
  block: blockRouter,
  tx: transactionRouter,
  blob: blobRouter,
  search: searchRouter,
  stats: statsRouter,
  indexer: indexerRouter,
  syncState: blockchainSyncStateRouter,
  blobStoragesState: blobStoragesStateRouter,
  healthcheck: publicProcedure
    .meta({
      openapi: {
        method: "GET",
        path: "/healthcheck",
        summary: "connection healthcheck.",
        tags: ["system"],
      },
    })
    .input(z.void())
    .output(z.string())
    .query(() => "yay!"),
});

// export type definition of API
export type AppRouter = typeof appRouter;
