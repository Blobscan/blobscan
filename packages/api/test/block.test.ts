/* eslint-disable @typescript-eslint/ban-ts-comment */
import type { inferProcedureInput } from "@trpc/server";
import { TRPCError } from "@trpc/server";
import { beforeAll, describe, expect, it } from "vitest";

import { fixtures } from "@blobscan/test";

import type { TRPCContext } from "../src";
import type { AppRouter } from "../src/app-router";
import { appRouter } from "../src/app-router";
import { createTestContext, runPaginationTestsSuite } from "./helpers";

type Input = inferProcedureInput<AppRouter["block"]["getByBlockIdFull"]>;

describe("Block router", async () => {
  let caller: ReturnType<typeof appRouter.createCaller>;
  let ctx: TRPCContext;

  beforeAll(async () => {
    ctx = await createTestContext();
    caller = appRouter.createCaller(ctx);
  });

  describe.each([{ functionName: "getAll" }, { functionName: "getAllFull" }])(
    "$functionName",
    ({ functionName }) => {
      runPaginationTestsSuite("block", (paginationInput) =>
        // @ts-ignore
        caller.block[functionName](paginationInput).then(({ blocks }) => blocks)
      );

      it("should the total number of blocks correctly", async () => {
        const expectedTotalBlocks = fixtures.blocks.length;

        await ctx.prisma.blockOverallStats.populate();
        await caller.block.getAllFull();

        // @ts-ignore
        const { totalBlocks } = await caller.block[functionName]();

        expect(totalBlocks).toBe(expectedTotalBlocks);
      });
    }
  );

  describe.each([
    { functionName: "getByBlockId" },
    { functionName: "getByBlockIdFull" },
  ])("$functionName", ({ functionName }) => {
    it("should get a block by hash", async () => {
      const input: Input = {
        id: "blockHash001",
      };

      const result = await caller.block[
        functionName as keyof typeof caller.block
      ](input);
      expect(result).toMatchSnapshot();
    });

    it("should fail when trying to get a block with a non-existent hash", async () => {
      await expect(
        caller.block[functionName as keyof typeof caller.block]({
          id: "nonExistingHash",
        })
      ).rejects.toThrow(
        new TRPCError({
          code: "NOT_FOUND",
          message: `No block with number, slot or hash 'nonExistingHash'.`,
        })
      );
    });

    it("should get a block by slot", async () => {
      const input: Input = {
        id: "101",
      };

      const result = await caller.block[
        functionName as keyof typeof caller.block
      ](input);
      expect(result).toMatchSnapshot();
    });

    it("should fail when trying to get a block with a non-existent slot", async () => {
      await expect(
        caller.block[functionName as keyof typeof caller.block]({
          id: "nonExistingSlot",
        })
      ).rejects.toThrow(
        new TRPCError({
          code: "NOT_FOUND",
          message: `No block with number, slot or hash 'nonExistingSlot'.`,
        })
      );
    });

    it("should get a block by block number", async () => {
      type Input = inferProcedureInput<AppRouter["block"]["getByBlockIdFull"]>;
      const input: Input = {
        id: "1002",
      };

      const result = await caller.block[
        functionName as keyof typeof caller.block
      ](input);
      expect(result).toMatchSnapshot();
    });

    it("should fail when trying to get a block with a non-existent block number", async () => {
      await expect(
        caller.block[functionName as keyof typeof caller.block]({
          id: "9999",
        })
      ).rejects.toThrow(
        new TRPCError({
          code: "NOT_FOUND",
          message: `No block with number, slot or hash '9999'.`,
        })
      );
    });
  });
});
