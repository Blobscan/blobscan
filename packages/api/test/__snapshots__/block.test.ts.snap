// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Block router > 'getAll' > when getting paginated block results > should default to the first page when no page was specified 1`] = `
[
  {
    "blobAsCalldataGasUsed": "255000",
    "blobGasPrice": "22",
    "blobGasUsed": "5500000",
    "blobs": [
      "blobHash001",
    ],
    "excessBlobGas": "15000",
    "hash": "blockHash008",
    "number": 1008,
    "slot": 108,
    "timestamp": 1693497600,
    "totalBlobSize": 1000,
    "transactions": [
      "txHash016",
    ],
  },
  {
    "blobAsCalldataGasUsed": "255000",
    "blobGasPrice": "22",
    "blobGasUsed": "5500000",
    "blobs": [
      "blobHash004",
      "blobHash004",
      "blobHash004",
    ],
    "excessBlobGas": "15000",
    "hash": "blockHash007",
    "number": 1007,
    "slot": 107,
    "timestamp": 1693490400,
    "totalBlobSize": 3900,
    "transactions": [
      "txHash015",
    ],
  },
]
`;

exports[`Block router > 'getAll' > when getting paginated block results > should return the correct number of results 1`] = `
[
  {
    "blobAsCalldataGasUsed": "255000",
    "blobGasPrice": "22",
    "blobGasUsed": "5500000",
    "blobs": [
      "blobHash001",
    ],
    "excessBlobGas": "15000",
    "hash": "blockHash008",
    "number": 1008,
    "slot": 108,
    "timestamp": 1693497600,
    "totalBlobSize": 1000,
    "transactions": [
      "txHash016",
    ],
  },
  {
    "blobAsCalldataGasUsed": "255000",
    "blobGasPrice": "22",
    "blobGasUsed": "5500000",
    "blobs": [
      "blobHash004",
      "blobHash004",
      "blobHash004",
    ],
    "excessBlobGas": "15000",
    "hash": "blockHash007",
    "number": 1007,
    "slot": 107,
    "timestamp": 1693490400,
    "totalBlobSize": 3900,
    "transactions": [
      "txHash015",
    ],
  },
]
`;

exports[`Block router > 'getAllFull' > when getting paginated block results > should default to the first page when no page was specified 1`] = `
[
  {
    "blobAsCalldataGasUsed": "255000",
    "blobGasPrice": "22",
    "blobGasUsed": "5500000",
    "excessBlobGas": "15000",
    "hash": "blockHash008",
    "number": 1008,
    "slot": 108,
    "timestamp": 1693497600,
    "totalBlobSize": 1000,
    "transactions": [
      {
        "blobs": [
          {
            "blob": {
              "size": 1000,
            },
            "blobHash": "blobHash001",
            "index": 0,
          },
        ],
        "fromId": "address2",
        "hash": "txHash016",
        "toId": "address4",
      },
    ],
  },
  {
    "blobAsCalldataGasUsed": "255000",
    "blobGasPrice": "22",
    "blobGasUsed": "5500000",
    "excessBlobGas": "15000",
    "hash": "blockHash007",
    "number": 1007,
    "slot": 107,
    "timestamp": 1693490400,
    "totalBlobSize": 3900,
    "transactions": [
      {
        "blobs": [
          {
            "blob": {
              "size": 1300,
            },
            "blobHash": "blobHash004",
            "index": 0,
          },
          {
            "blob": {
              "size": 1300,
            },
            "blobHash": "blobHash004",
            "index": 1,
          },
          {
            "blob": {
              "size": 1300,
            },
            "blobHash": "blobHash004",
            "index": 2,
          },
        ],
        "fromId": "address5",
        "hash": "txHash015",
        "toId": "address4",
      },
    ],
  },
]
`;

exports[`Block router > 'getAllFull' > when getting paginated block results > should return the correct number of results 1`] = `
[
  {
    "blobAsCalldataGasUsed": "255000",
    "blobGasPrice": "22",
    "blobGasUsed": "5500000",
    "excessBlobGas": "15000",
    "hash": "blockHash008",
    "number": 1008,
    "slot": 108,
    "timestamp": 1693497600,
    "totalBlobSize": 1000,
    "transactions": [
      {
        "blobs": [
          {
            "blob": {
              "size": 1000,
            },
            "blobHash": "blobHash001",
            "index": 0,
          },
        ],
        "fromId": "address2",
        "hash": "txHash016",
        "toId": "address4",
      },
    ],
  },
  {
    "blobAsCalldataGasUsed": "255000",
    "blobGasPrice": "22",
    "blobGasUsed": "5500000",
    "excessBlobGas": "15000",
    "hash": "blockHash007",
    "number": 1007,
    "slot": 107,
    "timestamp": 1693490400,
    "totalBlobSize": 3900,
    "transactions": [
      {
        "blobs": [
          {
            "blob": {
              "size": 1300,
            },
            "blobHash": "blobHash004",
            "index": 0,
          },
          {
            "blob": {
              "size": 1300,
            },
            "blobHash": "blobHash004",
            "index": 1,
          },
          {
            "blob": {
              "size": 1300,
            },
            "blobHash": "blobHash004",
            "index": 2,
          },
        ],
        "fromId": "address5",
        "hash": "txHash015",
        "toId": "address4",
      },
    ],
  },
]
`;

exports[`Block router > 'getByBlockId' > should get a block by block number 1`] = `
{
  "blobAsCalldataGasUsed": "255000",
  "blobGasPrice": "22",
  "blobGasUsed": "5500000",
  "blobs": [
    "blobHash001",
    "blobHash002",
    "blobHash003",
    "blobHash004",
    "blobHash005",
    "blobHash006",
  ],
  "excessBlobGas": "15000",
  "hash": "blockHash002",
  "number": 1002,
  "slot": 102,
  "timestamp": 1683720000,
  "totalBlobSize": 7500,
  "transactions": [
    "txHash004",
  ],
}
`;

exports[`Block router > 'getByBlockId' > should get a block by hash 1`] = `
{
  "blobAsCalldataGasUsed": "2042780",
  "blobGasPrice": "22",
  "blobGasUsed": "786432",
  "blobs": [
    "blobHash001",
    "blobHash002",
    "blobHash003",
    "blobHash001",
    "blobHash001",
    "blobHash002",
  ],
  "excessBlobGas": "15000",
  "hash": "blockHash001",
  "number": 1001,
  "slot": 101,
  "timestamp": 1665921600,
  "totalBlobSize": 6400,
  "transactions": [
    "txHash001",
    "txHash002",
    "txHash003",
  ],
}
`;

exports[`Block router > 'getByBlockId' > should get a block by slot 1`] = `
{
  "blobAsCalldataGasUsed": "2042780",
  "blobGasPrice": "22",
  "blobGasUsed": "786432",
  "blobs": [
    "blobHash001",
    "blobHash002",
    "blobHash003",
    "blobHash001",
    "blobHash001",
    "blobHash002",
  ],
  "excessBlobGas": "15000",
  "hash": "blockHash001",
  "number": 1001,
  "slot": 101,
  "timestamp": 1665921600,
  "totalBlobSize": 6400,
  "transactions": [
    "txHash001",
    "txHash002",
    "txHash003",
  ],
}
`;

exports[`Block router > 'getByBlockIdFull' > should get a block by block number 1`] = `
{
  "blobAsCalldataGasUsed": "255000",
  "blobGasPrice": "22",
  "blobGasUsed": "5500000",
  "excessBlobGas": "15000",
  "hash": "blockHash002",
  "number": 1002,
  "slot": 102,
  "timestamp": 1683720000,
  "totalBlobSize": 7500,
  "transactions": [
    {
      "blobs": [
        {
          "blob": {
            "size": 1000,
          },
          "blobHash": "blobHash001",
          "index": 0,
        },
        {
          "blob": {
            "size": 1100,
          },
          "blobHash": "blobHash002",
          "index": 1,
        },
        {
          "blob": {
            "size": 1200,
          },
          "blobHash": "blobHash003",
          "index": 2,
        },
        {
          "blob": {
            "size": 1300,
          },
          "blobHash": "blobHash004",
          "index": 3,
        },
        {
          "blob": {
            "size": 1400,
          },
          "blobHash": "blobHash005",
          "index": 4,
        },
        {
          "blob": {
            "size": 1500,
          },
          "blobHash": "blobHash006",
          "index": 5,
        },
      ],
      "fromId": "address1",
      "hash": "txHash004",
      "toId": "address3",
    },
  ],
}
`;

exports[`Block router > 'getByBlockIdFull' > should get a block by hash 1`] = `
{
  "blobAsCalldataGasUsed": "2042780",
  "blobGasPrice": "22",
  "blobGasUsed": "786432",
  "excessBlobGas": "15000",
  "hash": "blockHash001",
  "number": 1001,
  "slot": 101,
  "timestamp": 1665921600,
  "totalBlobSize": 6400,
  "transactions": [
    {
      "blobs": [
        {
          "blob": {
            "size": 1000,
          },
          "blobHash": "blobHash001",
          "index": 0,
        },
        {
          "blob": {
            "size": 1100,
          },
          "blobHash": "blobHash002",
          "index": 1,
        },
        {
          "blob": {
            "size": 1200,
          },
          "blobHash": "blobHash003",
          "index": 2,
        },
      ],
      "fromId": "address1",
      "hash": "txHash001",
      "toId": "address2",
    },
    {
      "blobs": [
        {
          "blob": {
            "size": 1000,
          },
          "blobHash": "blobHash001",
          "index": 0,
        },
      ],
      "fromId": "address3",
      "hash": "txHash002",
      "toId": "address4",
    },
    {
      "blobs": [
        {
          "blob": {
            "size": 1000,
          },
          "blobHash": "blobHash001",
          "index": 0,
        },
        {
          "blob": {
            "size": 1100,
          },
          "blobHash": "blobHash002",
          "index": 1,
        },
      ],
      "fromId": "address5",
      "hash": "txHash003",
      "toId": "address6",
    },
  ],
}
`;

exports[`Block router > 'getByBlockIdFull' > should get a block by slot 1`] = `
{
  "blobAsCalldataGasUsed": "2042780",
  "blobGasPrice": "22",
  "blobGasUsed": "786432",
  "excessBlobGas": "15000",
  "hash": "blockHash001",
  "number": 1001,
  "slot": 101,
  "timestamp": 1665921600,
  "totalBlobSize": 6400,
  "transactions": [
    {
      "blobs": [
        {
          "blob": {
            "size": 1000,
          },
          "blobHash": "blobHash001",
          "index": 0,
        },
        {
          "blob": {
            "size": 1100,
          },
          "blobHash": "blobHash002",
          "index": 1,
        },
        {
          "blob": {
            "size": 1200,
          },
          "blobHash": "blobHash003",
          "index": 2,
        },
      ],
      "fromId": "address1",
      "hash": "txHash001",
      "toId": "address2",
    },
    {
      "blobs": [
        {
          "blob": {
            "size": 1000,
          },
          "blobHash": "blobHash001",
          "index": 0,
        },
      ],
      "fromId": "address3",
      "hash": "txHash002",
      "toId": "address4",
    },
    {
      "blobs": [
        {
          "blob": {
            "size": 1000,
          },
          "blobHash": "blobHash001",
          "index": 0,
        },
        {
          "blob": {
            "size": 1100,
          },
          "blobHash": "blobHash002",
          "index": 1,
        },
      ],
      "fromId": "address5",
      "hash": "txHash003",
      "toId": "address6",
    },
  ],
}
`;
